nano/node/repcrawler.cpp

```diff
@@ -1,370 +1,524 @@
+{
+	debug_assert (!mutex.try_lock ());
+	debug_assert (lock.owns_lock ());
+	debug_assert (!responses.empty ()); // Should be checked before calling this function
+
+	decltype (responses) responses_l{ responses.capacity () };
+	responses_l.swap (responses);
+
+	lock.unlock ();
 
 	// normally the rep_crawler only tracks principal reps but it can be made to track
 	// reps with less weight by setting rep_crawler_weight_minimum to a low value
-	auto minimum = std::min (node.minimum_principal_weight (), node.config.rep_crawler_weight_minimum.number ());
+	auto const minimum = std::min (node.minimum_principal_weight (), node.config.rep_crawler_weight_minimum.number ());
 
-	for (auto const & i : responses_l)
+	// TODO: Is it really faster to repeatedly lock/unlock the mutex for each response?
+	for (auto const & response : responses_l)
 	{
-		auto & vote = i.second;
-		auto & channel = i.first;
-		debug_assert (channel != nullptr);
+		auto & vote = response.second;
+		auto & channel = response.first;
+		release_assert (vote != nullptr);
+		release_assert (channel != nullptr);
 
 		if (channel->get_type () == nano::transport::transport_type::loopback)
 		{
-			node.logger.debug (nano::log::type::repcrawler, "Ignoring vote from loopback channel: {}", channel->to_string ());
-
+			logger.debug (nano::log::type::rep_crawler, "Ignoring vote from loopback channel: {}", channel->to_string ());
 			continue;
 		}
 
-		nano::uint128_t rep_weight = node.ledger.weight (vote->account);
+		nano::uint128_t const rep_weight = node.ledger.weight (vote->account);
 		if (rep_weight < minimum)
 		{
-			node.logger.debug (nano::log::type::repcrawler, "Ignoring vote from account {} with too little voting weight: {}",
+			logger.debug (nano::log::type::rep_crawler, "Ignoring vote from account {} with too little voting weight: {}",
 			vote->account.to_account (),
 			nano::util::to_str (rep_weight));
-
 			continue;
 		}
 
 		// temporary data used for logging after dropping the lock
-		auto inserted = false;
-		auto updated = false;
+		bool inserted = false;
+		bool updated = false;
 		std::shared_ptr<nano::transport::channel> prev_channel;
 
-		nano::unique_lock<nano::mutex> lock{ probable_reps_mutex };
+		lock.lock ();
 
-		auto existing (probable_reps.find (vote->account));
-		if (existing != probable_reps.end ())
+		if (auto existing = reps.find (vote->account); existing != reps.end ())
 		{
-			probable_reps.modify (existing, [rep_weight, &updated, &vote, &channel, &prev_channel] (nano::representative & info) {
-				info.last_response = std::chrono::steady_clock::now ();
+			reps.modify (existing, [rep_weight, &updated, &vote, &channel, &prev_channel] (rep_entry & rep) {
+				rep.last_response = std::chrono::steady_clock::now ();
 
 				// Update if representative channel was changed
-				if (info.channel->get_endpoint () != channel->get_endpoint ())
+				if (rep.channel->get_endpoint () != channel->get_endpoint ())
```

Comment:
Technically, it would be better to compare channel rather than endpoints.
It is possible two have two channels with the same remote endpoint, if the other computer suddenly reboots.
This will also work because eventually the old channel should timeout but it will take longer to notice.

Reply:
We initiate two channels between peers, one in, one out. Isn't comparing channels going to lead to excessive channel switching? I assume the original author of the code used this comparisons style to fix this problem.

Reply:
Hmmm, maybe that is the reason. OK.